Add support for WooCommerce Payments ad banner on extensions screen's "Featured" section

andfinally opened the pull request:
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:
We want to be able to show a banner on the wc-admin featured extensions page, to highlight the benefits WooCommerce Pay offers to store owners and customers.

- Add support for a banner on the featured extensions section.
- Add CSS rules for the banner to `admin.scss`.

### WooCommerce.com issue

This change is related to 9860-gh-Automattic/woocommerce.com.

### How to test the changes in this Pull Request:

1. Check the branch out locally and build assets.
2. If the update to the WooCommerce.com featured endpoint  in 10179-gh-Automattic/woocommerce.com hasn't been merged yet, you'll need to modify the featured endpoint locally to see the ad. Copy the JSON from [here](293e7-pb/) and make it available at some publicly-accessible URL, like a GitHub gist. Go to `includes/admin/class-wc-admin-addons.php` and replace

```php
$raw_featured = wp_safe_remote_get(
	'https://woocommerce.com/wp-json/wccom-extensions/1.0/featured',
	array(
		'headers'    => $headers,
		'user-agent' => 'WooCommerce Addons Page',
	)
);
```

with

```php
$raw_featured = wp_safe_remote_get(
	[ URL for the file ],
);
```

You may need to flush the object cache and featured transient to see the changes:

```
wp cache flush
wp transient delete wc_addons_featured
```
3. To see the ad, WCPay should not be active on your test site, and your location should be in the US, Canada, UK, Ireland, Australia or New Zealand. If you're not in one of these countries, you can cheat by commenting out lines 793-795 of `includes/admin/class-wc-admin-addons.php`:

```php
if ( isset( $item->geowhitelist ) && ! in_array( $location, $item->geowhitelist, true ) ) {
    return false;
}
```

4. Go to the WooCommerce > Extensions page in your test site. You should see the "featured" section (`/wp-admin/admin.php?page=wc-addons&section=_featured`). Check that the ad appears as expected, and the links work.
5. Install and activate WCPay. The ad should not appear.
6. For bonus points, view the screen from an unsupported location, or hack the `show_extension` method so it doesn't like your country. The ad should not appear.

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [x] Have you successfully run tests with your changes locally?

<!-- Mark completed items with an [x] -->

### Changelog entry

New - Feature WooCommerce Payments in the extensions store for stores in the US, Canada, UK, Ireland, Australia and New Zealand.

### Screenshots

<img width="300" src="https://user-images.githubusercontent.com/1647564/117308582-993b2c80-ae79-11eb-9dd9-02046a67494d.png"/>

<img width="500" src="https://user-images.githubusercontent.com/1647564/117308696-b3750a80-ae79-11eb-98fe-a4981495f0df.png"/>

andfinally committed:
Added `output_ad_block` method to `WC_Admin_Addons` and CSS rules to render a WooCommerce Payments ad banner at the top of the "Featured" section of the WC Core addons page, `wp-admin/admin.php?page=wc-addons&section=_featured`. This is to fulfil the requirements of WooCommerce.com issue https://github.com/Automattic/woocommerce.com/issues/9860.

We output the block if

- The user has permissions to install plugins.
- Their location is in one of the countries identified in a `geowhitelist`.
- WooCommerce Payments is not already active.

andfinally committed:
Added comment to new `output_ad_block` method. Removed differences introduced by PhpStorm autoformat.

andfinally committed:
Added docblock comment for `$section` param.

andfinally committed:
Fixing doc comment for new method.

andfinally committed:
Changed block name and CSS classnames to the more generic "promotion block".

vedanshujain reviewed (dismissed):
Let some minor comments, otherwise testing well and looks good! Great work!

andfinally committed:
Addressing feedback. Removed `!important` from `.addons-button-expandable` rule. Renamed `geowhitelist` and `geoblacklist` to `geo_allow_list` and `geo_block_list`.

andfinally commented:
Many thanks for taking a look @vedanshujain, I've hopefully addressed your points. I found the `!important` wasn't necessary after all. I saved an updated version of the test json in 293ca-pb/.

andfinally commented on includes/admin/class-wc-admin-addons.php:
```diff
@@ -534,6 +534,69 @@ class="addons-img"
+
+		$section_object = (object) $section;
+
+		if ( ! empty( $section_object->geowhitelist ) ) {
```
Thanks, changed in de331ad8935d01eafc910d84fc4464469cfd1dac.

andfinally replied:
Reverted in 6bb1ec73a4757cf38ddb1893a06515ccb6263b57.

andfinally commented on assets/css/admin.scss:
```diff
@@ -364,6 +401,12 @@
+	.addons-button-expandable {
+		display: inline-block;
+		padding: 0 16px;
+		width: auto !important;
```
I've removed this in de331ad8935d01eafc910d84fc4464469cfd1dac – the `!important` wasn't needed after all.

vedanshujain commented on includes/admin/class-wc-admin-addons.php:
```diff
@@ -722,11 +788,11 @@ public static function install_woocommerce_payments_addon( $section = '_featured
 	public static function show_extension( $item ) {
 		$location = WC()->countries->get_base_country();
-		if ( isset( $item->geowhitelist ) && ! in_array( $location, $item->geowhitelist, true ) ) {
+		if ( isset( $item->geo_allow_list ) && ! in_array( $location, $item->geo_allow_list, true ) ) {
```
hi, sorry, i did not realize that these naming are also being used here as well, which is being used for other items as well and not just wc pay. we should revert this change, because we don't know if all the other promotional extensions have changed the wording as well, and they have not then these wordings will not work for them anymore.

vedanshujain commented on includes/admin/class-wc-admin-addons.php:
```diff
@@ -534,6 +534,69 @@ class="addons-img"
+
+		$section_object = (object) $section;
+
+		if ( ! empty( $section_object->geo_allow_list ) ) {
```
should not we do the same for block list as well? as of now blocklist won't won't work

andfinally replied:
Good point – added in 7202f19631e474d1adc93d83acc1499bab0b6b3d.

vedanshujain reviewed (commented):
Looks like we may want to revert one of the feedback we addressed because of something I missed earlier. Sorry about that!

andfinally committed:
Reverted renaming of `geowhitelist` and `geoblacklist`, in case they are still present in other endpoints.

andfinally committed:
Added support for `geoblacklist` in `output_promotion_block`.

andfinally commented:
Thanks Vedanshu, I've reverted the rename. I did search for those names in the WCCOM codebase and didn't find them in any other endpoints, so it would probably have been OK. But it's good to be cautious.

vedanshujain reviewed (approved):
LGTM!
